### PR TITLE
[11.x] Add an option to skip certain migration files

### PR DESCRIPTION
**Pull request by tkoop**
This is a relatively small code change. It only adds an extra option to migrations. It doesn't change any existing behavior. But it gives us a good tool in testing migrations.

It simply adds the option to skip certain migration files when running a migration, using the `--skip` option.

Why is this needed? Imagine this situation:

You are going to make a complex change to the database, so your new migration file is long and complicated. And to make sure it will do its thing correctly, you want to manually make some changes to the database first, or perhaps even import live data, just to make sure it will migrate correctly.

With the "skip" option, you can do this:
1. `php artisan migrate:fresh --seed --skip="2024_04_27_complex_migration.php"`
2. Manually import your live or complex data.
3. `php artisan migrate`

Without the skip option, how else could you do this? Here are some alternatives and why they're not as good as skipping.

- You could add a line at the top of the migration file that seeds the database with a custom seeder, but that means you would have to create a seeder just for this purpose. You would prefer to import the new data manually. It also means that you have to remove this line from the migration file when you are done testing. It would be better to keep the migration file exactly as is, without adding or removing lines.
- You could use git to shuffle branches around, to hide the new migration file, then checkout the new branch to show it. But this requires knowledge of git and is a complex solution. Skipping is a lot simpler.
- You could migrate everything, then roll back that last migration, then add the data, then migrate again. The problem with this is that the migration is so complex, you would have to put a lot of time into the "down" code, which will never actually get used otherwise, and which might itself be wrong. Skipping would mean you wouldn't have to write "down" code.

So in conclusion, I humble submit this pull request for your consideration. It won't change any existing features, and it will make everyone's life easier.

Thank you.

### CODE REVIEW

**Comment by devajmeireles** [comment]
Why not implement a selector via Laravel Prompts to improve the ease of use of this?

**Comment by tkoop** [comment]
> Why not implement a selector via Laravel Prompts to improve the ease of use of this?

Do you mean the user should be prompted for which migration to skip instead of typing out the file name manually? If so, this sounds like a great idea. However, that would take a little more effort, so I don't mind if this merge request gets merged first.

**Comment by devajmeireles** [comment]
> > Why not implement a selector via Laravel Prompts to improve the ease of use of this?
> 
> Do you mean the user should be prompted for which migration to skip instead of typing out the file name manually? If so, this sounds like a great idea. However, that would take a little more effort, so I don't mind if this merge request gets merged first.

The idea behind using prompts is to load all current migration files and display them in a [multi-selection prompt](https://laravel.com/docs/11.x/prompts#multiselect), allowing us to select multiples to ignore.

**Comment by rodrigopedra** [comment]
> Without the skip option, how else could you do this?

```bash
mkdir -p ./database/migrations/skip
mv ./database/migrations/2024_04_27_complex_migration.php ./database/migrations/skip/
php artisan migrate
# do complex data operations
mv ./database/migrations/skip/2024_04_27_complex_migration.php ./database/migrations/
php artisan migrate
```

**Comment by joshmanders** [comment]
Could also automatically prompt with multiselect if `--skip` is provided but without any migrations as the value.

**Comment by driesvints** [comment]
> You could add a line at the top of the migration file that seeds the database with a custom seeder,

This isn't what migrations are intended for and should never be done.

> You could use git to shuffle branches around, to hide the new migration file,

If you switch branches between versions with different migrations, you should re-migrate your database.

> You could migrate everything, then roll back that last migration, then add the data, then migrate again.

Just use `php artisan migrate:fresh`. As for production, honestly I've never ever used a down migration in my entire career nor do I believe it's widely used.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Comment by tkoop** [comment]
@taylorotwell So in the above scenario, what do you recommend a developer should do?

**Comment by tkoop** [comment]
@taylorotwell So what's the best way to handle the above scenario?

**Comment by melokullc** [comment]
I am kind of in the ballpark with @rodrigopedra on this one. Normally, I would just move the migration into a migrated folder, run migrations, then either keep the skipped migration in the migrated folder or dump it back into migrations.

**Comment by Muetze42** [comment]
@tkoop 

"you want to manually make some changes to the database first"
Why should you do that? There are migration and tests for that too?

"You would prefer to import the new data manually"
Is't possible to export databases without structure?

- Delete the migration files to skip or remove the php extension for this files
- Run migrate command
- Restore or add the php exstension for the files and
- Optional: Add the migrations to the migrate table
(Just as rodrigopedra described it.)

Otherwise, it is as driesvints has already written: 
"This isn't what migrations are intended for and should never be done."
